sql: crash when distributing subquery

Set up a 2+ node cluster. Then:

```
CREATE TABLE ab (a INT, b INT);
INSERT INTO ab VALUES (1, 1), (1, 3), (2, 2);
SET CLUSTER SETTING kv.range_merge.queue_enabled = false;
ALTER TABLE ab SPLIT AT VALUES (2);
ALTER TABLE ab EXPERIMENTAL_RELOCATE VALUES (ARRAY[1], 1), (ARRAY[2], 2);
SHOW EXPERIMENTAL_RANGES FROM TABLE ab;
```

output of the last one should be:
```
  start_key | end_key | range_id | replicas | lease_holder
+-----------+---------+----------+----------+--------------+
  NULL      | /2      |       21 | {1}      |            1
  /2        | NULL    |       22 | {2}      |            2
(2 rows)
```

Then running
```
SET OPTIMIZER=OFF; SET DISTSQL=ON; SELECT ARRAY(SELECT a FROM ab ORDER BY b);
```
crashes with:
```
panic: invalid row length 1, expected 2 [recovered]
	panic: panic while executing 1 statements: SELECT ARRAY (SELECT _ FROM _ ORDER BY _); caused by invalid row length 1, expected 2

goroutine 622 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc4293bd300, 0x31e2f80, 0xc42b2d8940, 0x27abb60, 0xc420340d30)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:647 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc4293bd300, 0x31e2f80, 0xc42b2d8940)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:388 +0x61
panic(0x27abb60, 0xc420340d30)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowContainer).AddRow(0xc428ee34d0, 0x31e3040, 0xc420ca5320, 0xc420340d20, 0x1, 0x1, 0x6a7c07, 0x10, 0x2ab3ee0, 0x1, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/row_container.go:282 +0x46c
github.com/cockroachdb/cockroach/pkg/sql.(*RowResultWriter).AddRow(0xc42ae793e0, 0x31e3040, 0xc420ca5320, 0xc420340d20, 0x1, 0x1, 0x0, 0x0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_plan_csv.go:116 +0x69
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).Push(0xc42989fb80, 0xc4292f9c28, 0x2, 0x2, 0x0, 0x2)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:557 +0x8bc
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*accountClearingRowReceiver).Push(0xc4292bdd10, 0xc4292f9c28, 0x2, 0x2, 0x0, 0x2)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:279 +0x83
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*copyingRowReceiver).Push(0xc4292f9c00, 0xc42bd173e0, 0x2, 0x2, 0x0, 0xc400000000)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:673 +0x191
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x31e3040, 0xc4292bdb90, 0x31ee040, 0xc429384d80, 0x31d54c0, 0xc4292f9c00)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:177 +0x5c
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc429384d80, 0x31e3040, 0xc4292bdb90, 0x0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:768 +0x98
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).StartSync(0xc42952ae00, 0x31e3040, 0xc4292bdb90, 0x2da0790, 0xc42bd07a40, 0x31d23c0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:630 +0x191
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc42093b800, 0xc428e870e0, 0xc428ee2e10, 0xc42cb2c8e8, 0xc42989fb80, 0xc4295232c0, 0x0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:261 +0x868
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planAndRunSubquery(0xc42093b800, 0x31e3040, 0xc420ca5320, 0x0, 0xc420ca44b0, 0x3, 0x1, 0x31d3500, 0xc42989c070, 0x0, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:715 +0x858
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunSubqueries(0xc42093b800, 0x31e3040, 0xc420ca5320, 0xc4293bd728, 0xc420ca57d0, 0xc429201c00, 0x1, 0x1, 0xc42972e000, 0xc4293bd401, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:627 +0x18b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc4293bd300, 0x31e3040, 0xc420ca5320, 0xc4293bd728, 0x3, 0x7f80fff542f0, 0xc428ee2fc0, 0x1, 0x0, 0x0)
```

This is the plan for the subquery (used the COCKROACH_DISTSQL_LOG_PLAN flag):
https://cockroachdb.github.io/distsqlplan/decode.html#eJyskbFqwzAQhvc-Rbm1KrFkd9HkNUtS0m7Fg2IdweD4xEmGluB3L7YKqdtEdSGj7u677_h1go4sbswRPeg3kCBAQSXAMdXoPfFYjkNr-w46E9B0rg9juRJQEyPoE4QmtAgaXs2-xR0ai7zKQIDFYJp2Wu24ORr-KM0eBGz7oO9LKUoF1SCA-nDe6YM5IGg5iOXeF-KAvFJzZakelrnUVddZQWyR0f42VMOFgzb0SG5VzKav2fOZXS5PWN4y4T-8XwnnN0_4gmuH3lHncVF62Rg_2gPG7_LUc43PTPWkic_txE0Fiz7EroyPdRdb44HfYZmE1QyWP2GVhJ_S5jwJF2m4-NfZ1XD3GQAA__8MJk2B

What is happening here is that, even though the subquery projects away `b`, the distsql planner is forced to keep `b` around in order to merge the streams correctly. The end result is that we get two columns instead of one.

The `Run` code effects the final projection using `plan.PlanToStreamColMap` (as `resultToStreamColMap`). Unfortunately the `planAndRunSubquery` code does not take that into account when setting up the container. It should use the result types from `planColumns(subqueryPlan.plan)` instead of `subqueryPhysPlan.ResultTypes` (or alternatively, it should apply the `PlanToStreamColMap` projection to `ResultTypes`).

CC @jordanlewis

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -699,12 +699,24 @@ func (dsp *DistSQLPlanner) planAndRunSubquery(
 	// TODO(arjun): #28264: We set up a row container, wrap it in a row
 	// receiver, and use it and serialize the results of the subquery. The type
 	// of the results stored in the container depends on the type of the subquery.
-	typ := sqlbase.ColTypeInfoFromColTypes(subqueryPhysPlan.ResultTypes)
 	subqueryRecv := recv.clone()
+	var typ sqlbase.ColTypeInfo
 	var rows *sqlbase.RowContainer
 	if subqueryPlan.execMode == distsqlrun.SubqueryExecModeExists {
 		subqueryRecv.noColsRequired = true
 		typ = sqlbase.ColTypeInfoFromColTypes([]sqlbase.ColumnType{})
+	} else {
+		// Apply the PlanToStreamColMap projection to the ResultTypes to get the
+		// final set of output types for the subquery. The reason this is necessary
+		// is that the output schema of a query sometimes contains columns necessary
+		// to merge the streams, but that aren't required by the final output of the
+		// query. These get projected out, so we need to similarly adjust the
+		// expected result types of the subquery here.
+		colTypes := make([]sqlbase.ColumnType, len(subqueryPhysPlan.PlanToStreamColMap))
+		for i, resIdx := range subqueryPhysPlan.PlanToStreamColMap {
+			colTypes[i] = subqueryPhysPlan.ResultTypes[resIdx]
+		}
+		typ = sqlbase.ColTypeInfoFromColTypes(colTypes)
 	}
 	rows = sqlbase.NewRowContainer(subqueryMemAccount, typ, 0)
 	defer rows.Close(evalCtx.Ctx())

--- a/pkg/sql/logictest/testdata/logic_test/distsql_subquery
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_subquery
@@ -1,0 +1,33 @@
+# LogicTest: 5node-dist
+
+# TODO(radu): re-add 5node-dist-opt when #32648 is fixed.
+
+# Regression test for #32652: make sure subqueries that have extra columns for
+# stream merges don't crash when executed.
+
+statement ok
+CREATE TABLE ab (a INT, b INT)
+
+statement ok
+INSERT INTO ab VALUES (1, 1), (1, 3), (2, 2)
+
+statement ok
+SET CLUSTER SETTING kv.range_merge.queue_enabled = false
+
+statement ok
+ALTER TABLE ab SPLIT AT VALUES (2)
+
+statement ok
+ALTER TABLE ab EXPERIMENTAL_RELOCATE VALUES (ARRAY[1], 1), (ARRAY[2], 2)
+
+query TTITI colnames
+SHOW EXPERIMENTAL_RANGES FROM TABLE ab
+----
+start_key  end_key  range_id  replicas  lease_holder
+NULL       /2       1         {1}       1
+/2         NULL     2         {2}       2
+
+query T
+SELECT ARRAY(SELECT a FROM ab ORDER BY b)
+----
+{1,2,1}

